File: Java/Hard/2751 Robot Collisions.java
class Solution {
    public List<Integer> survivedRobotsHealths(int[] positions, int[] healths, String directions) {
        int n = positions.length;
        Integer[] indices = new Integer[n];
        for (int i = 0; i < n; i++) indices[i] = i;

        // Sort indices based on their positions
        Arrays.sort(indices, (a, b) -> Integer.compare(positions[a], positions[b]));

        Deque<Integer> stack = new ArrayDeque<>();
        for (int i : indices) {
            if (directions.charAt(i) == 'R') {
                stack.push(i);
            } else {
                // Robot moving Left: check for collisions with Right-moving robots in stack
                while (!stack.isEmpty() && healths[i] > 0) {
                    int top = stack.peek();
                    if (healths[top] < healths[i]) {
                        // Right-moving robot destroyed
                        healths[i]--;
                        healths[top] = 0;
                        stack.pop();
                    } else if (healths[top] > healths[i]) {
                        // Left-moving robot destroyed
                        healths[top]--;
                        healths[i] = 0;
                    } else {
                        // Both destroyed
                        healths[i] = 0;
                        healths[top] = 0;
                        stack.pop();
                    }
                }
            }
        }

        List<Integer> result = new ArrayList<>();
        for (int h : healths) {
            if (h > 0) result.add(h);
        }
        return result;
    }
}

File: Python/Hard/2751 Robot Collisions.py
class Solution:
    def survivedRobotsHealths(self, positions: List[int], healths: List[int], directions: str) -> List[int]:
        n = len(positions)
        # Create a list of indices and sort them based on their positions
        robots = sorted(range(n), key=lambda i: positions[i])
        
        stack = []  # To store indices of robots moving 'R'
        
        for i in robots:
            if directions[i] == 'R':
                stack.append(i)
            else:
                # Robot is moving 'L', check for collisions with 'R' robots in stack
                while stack and healths[i] > 0:
                    top_idx = stack[-1]
                    
                    if healths[top_idx] < healths[i]:
                        # Left robot wins, Right robot destroyed
                        healths[top_idx] = 0
                        healths[i] -= 1
                        stack.pop()
                    elif healths[top_idx] > healths[i]:
                        # Right robot wins, Left robot destroyed
                        healths[top_idx] -= 1
                        healths[i] = 0
                    else:
                        # Both destroyed
                        healths[top_idx] = 0
                        healths[i] = 0
                        stack.pop()
        
        # Return healths of survivors in their original input order
        return [h for h in healths if h > 0]
